client/question/services/EventHandlerService.js

```diff
@@ -19,18 +19,58 @@
 tie.factory('EventHandlerService', [
   '$http', 'ServerHandlerService', function($http, ServerHandlerService) {
 
+    /**
+     * Global object to keep track of the current batch of events to send.
+     * @type [Event]
+     */
+    var _currentEventBatch = [];
+
+    /**
+     * Submits the current batch of event data to TIE's backend.
+     *
+     */
+    var sendCurrentEventBatch = function() {
+      if (getCurrentEventBatchLength() === 0) {
+        // No point in sending an empty batch.
+        return null;
+      }
+      var data = {
+        events: _currentEventBatch,
+        timeSentToBackendMsec: (new Date()).getTime()
+      };
+      return $http.post('/ajax/event/send_event_batch', data).then(
+        function() {
+          _currentEventBatch.length = 0;
+        }, function() {
+          // Otherwise, we should do nothing.
+          // Since the call errored, don't dump the current batch.
```

Comment:
Just to check, do we have any guarantees that a backend error here implies that no event in the batch was recorded? Need to make sure we don't process a batch partially and then error, and that this constraint/requirement holds going forward. (Might want to make a note to that effect in the corresponding backend changelist.)

Reply:
Hm. That's a good point, but I don't think we have that constraint, currently. What we COULD do is just create all the objects and then put() them in a final push, rather than create and put them at the same time?

Reply:
Yes, that seems reasonable. We'll also need a comment in the relevant place to ensure that that state of affairs is preserved going forward.

Reply:
I'll try to find a place to add that in my current fix.